File: src/stages/stage3.ts
import path from "node:path";

import { runParallelLimited, coerceError } from "../utils.js";
import { runWithValidation } from "../agents/index.js";
import { CheckpointManager } from "../checkpoint.js";
import type { AuditConfig, EntryPoint, Module } from "../config.js";
import { getLogger } from "../logger.js";
import { parseEntryPoints } from "../parsing/stage3.js";
import { loadPrompt } from "../prompts.js";
import { validateStage3File } from "../validation/stage3.js";

const logger = getLogger("stage3");

function taskKey(module: Module): string {
  return `stage3:${module.id}`;
}

async function runModule(
  module: Module,
  config: AuditConfig,
  checkpoint: CheckpointManager,
  stage2Output: string,
): Promise<EntryPoint[]> {
  const key = taskKey(module);
  const resultDir = path.join(config.outputDir, "stage-3-details");
  const outputPath = path.join(resultDir, `${module.id}.md`);

  if (checkpoint.isComplete(key)) {
    logger.info("Stage 3: %s already complete, loading existing output.", module.id);
    return parseEntryPoints(outputPath, module.id);
  }

  const prompt = await loadPrompt("stage3.md", {
    stage2_output_path: stage2Output,
    result_dir: resultDir,
    module_id: module.id,
  });

  const { passed } = await runWithValidation({
    config,
    prompt,
    cwd: config.target,
    outputPath,
    validator: validateStage3File,
  });

  if (!passed) {
    logger.warning("Stage 3: %s validation did not fully pass.", module.id);
  }

  checkpoint.markComplete(key);
  const entryPoints = parseEntryPoints(outputPath, module.id);
  logger.info(
    "Stage 3: %s complete. Entry points: %s",
    module.id,
    entryPoints.map((entryPoint) => entryPoint.id).join(", "),
  );
  return entryPoints;
}

export async function runStage3(
  modules: Module[],
  config: AuditConfig,
  checkpoint: CheckpointManager,
): Promise<Record<string, EntryPoint[]>> {
  const stage2Output = path.join(config.outputDir, "stage-2-scope.md");
  const results = await runParallelLimited(modules, config.maxParallel, async (module) => {
    return await runModule(module, config, checkpoint, stage2Output);
  });

  const entryPointMap: Record<string, EntryPoint[]> = {};
  results.forEach((result, index) => {
    const module = modules[index];
    if (!module) {
      return;
    }
    if (result.status === "rejected") {
      logger.error("Stage 3: %s failed with exception: %s", module.id, coerceError(result.reason).message);
      entryPointMap[module.id] = [];
      return;
    }
    entryPointMap[module.id] = result.value;
  });

  return entryPointMap;
}

File: protocol_auditor/stages/stage3.py
from __future__ import annotations

import os
import re

from ..agent import run_agent
from ..checkpoint import CheckpointManager
from ..config import AnalysisUnit, AuditConfig
from ..logger import get_logger
from ..prompts import load_prompt
from ..utils import format_validation_issues, list_matching_files, run_parallel_limited
from ..validation.stage3 import validate_stage3_file

logger = get_logger("stage3")


def _task_key(unit: AnalysisUnit) -> str:
    return f"stage3:{unit.id}"


async def _run_unit(
    unit: AnalysisUnit,
    config: AuditConfig,
    checkpoint: CheckpointManager,
) -> list[str]:
    key = _task_key(unit)
    result_dir = os.path.join(config.output_dir, "stage-3-details")
    escaped_id = re.escape(unit.id)
    finding_pattern = re.compile(rf"^{escaped_id}-F-\d+\.json$")

    if checkpoint.is_complete(key):
        logger.info("Stage 3: %s already complete, skipping.", unit.id)
        return list_matching_files(result_dir, finding_pattern)

    prompt = load_prompt("stage3.md", {
        "au_file_path": unit.au_file_path,
        "result_dir": result_dir,
        "finding_prefix": unit.id,
    })

    await run_agent(prompt, config, cwd=config.target)

    finding_files = list_matching_files(result_dir, finding_pattern)
    for finding_file in finding_files:
        issues = validate_stage3_file(finding_file)
        if not issues:
            continue

        logger.warning("Stage 3: Validation failed for %s\n%s", finding_file, format_validation_issues(issues))
        repair_prompt = (
            f"The finding file at `{finding_file}` failed validation. "
            f"Please fix all issues listed below:\n\n```\n{format_validation_issues(issues)}\n```"
        )
        await run_agent(repair_prompt, config, cwd=config.target, max_turns=10)

        issues = validate_stage3_file(finding_file)
        if issues:
            logger.warning("Stage 3: Repair failed for %s\n%s", finding_file, format_validation_issues(issues))

    checkpoint.mark_complete(key)
    logger.info("Stage 3: %s complete. Findings: %s", unit.id, len(finding_files))
    return finding_files


async def run_stage3(
    units: list[AnalysisUnit],
    config: AuditConfig,
    checkpoint: CheckpointManager,
) -> list[str]:
    if not units:
        logger.warning("Stage 3: No analysis units to process.")
        return []

    results = await run_parallel_limited(
        units,
        config.max_parallel,
        lambda unit, _: _run_unit(unit, config, checkpoint),
    )

    all_finding_files: list[str] = []
    for i, (status, value, error) in enumerate(results):
        if i >= len(units):
            continue
        if status == "rejected":
            logger.error("Stage 3: %s failed: %s", units[i].id, error)
            continue
        if value:
            all_finding_files.extend(value)

    logger.info("Stage 3 complete. Total bug finding files: %s", len(all_finding_files))
    return all_finding_files
